Added a liberal `Chunk` class.

dkorolev opened the pull request.

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+// and that the length-plus-first byte, at index `[length()]`, is set to '\0'.
+//
+// `UniquePiece` is a `Piece` that assumes each distinct string is stored only once. This requirement makes
+// their underlying `const char*`-s testable for equality. Long live and rest in piece, C+=, we'll miss you.
```
`C+=` -> `C++`

dkorolev replied:
Yes, those guys. ("Guys", I'd assume, https://www.youtube.com/watch?v=h0YC3RJA-ec)

sompylasar replied:
Haven't you noticed the effect of what you drank in the morning? :smiling_imp:

dkorolev replied:
Да, торкнуло неплохо. :trollface: :rainbow:

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+// because every decent architect knows well what happens if `free()` is invoked a bit too prematurely in the
+// evolutionary process of a lifetime of an object. (If you don't believe in evolution, don't use this code.)
+//
+// Note 1: It's not impossible for the order of pointed to strings to actually be lexicographical. However,
```
Can't parse `for the order of pointed to strings`, it there any typo?

dkorolev replied:
Intentionally complex to read: for the order of the strings that are being pointed to.

sompylasar replied:
Thanks!

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+// evolutionary process of a lifetime of an object. (If you don't believe in evolution, don't use this code.)
+//
+// Note 1: It's not impossible for the order of pointed to strings to actually be lexicographical. However,
+//         in most cases, making it so would be quite a large amount of unnecessary work. C+=. Never forget.
```
`C+=` again?

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+//         which is noticeably more efficient than a "regular" lexicographical comparison in O(N).
+//         Yes, it moves the world forward, saves trees and stops, or at least delays, global warming.
+//         Thus, non-metaphorically, the description above makes total sense. Keep in mind though, that
+//         making sense did not help C+= survive. At same same time, although contradictory to most predominant
```
`C+=`?

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+  void CheckPrivilege(size_t i) const { assert(i <= N); }
+
+  char operator[](size_t i) const {
+    CheckPrivilege(i);  // Once is enough.
```
What do you mean by "once" here? It's not obvious without looking into `CheckPrivilege` that the program will exit with an assertion exception.

dkorolev replied:
It's those guys too ...

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+        hash += k * piece.S[i];
+      }
+      static_assert(sizeof(double) >= sizeof(size_t), "Suddenly, `reinterpet_cast<>` doesn't nail it.");
+      return *reinterpret_cast<size_t*>(&hash);
```
Just curious, is `double` laid out in memory the same way as `size_t` so it can be `reinterpret_cast`ed?

dkorolev replied:
While ANSI regulates how exactly `double` laid out, it's but, but it's not important -- the important thing is that this hash function returns the same `size_t` values for same strings, and, hopefully very likely, different values of `size_t` for different strings.

sompylasar replied:
OK.

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+    bool operator()(const Piece& lhs, const Piece& rhs) const { return lhs.LexicographicalCompare(rhs) == 0; }
+  };
+
+  typedef EqualityComparator Pride;  // Your favorite equality smiley here. Mine is *HAWAII*, because rainbow.
```
because gladiolus?

dkorolev replied:
Yes. That too.

sompylasar commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+struct PieceDB {
+  std::unordered_map<Piece, const UniquePiece*, Piece::HashFunction, Piece::EqualityComparator> map;
+
+  // Note that `operator()` requires a non-const `Piece&`. This is done to ensure no temporary `Piece` object
```
`operator[]`, not `()` is written below.

dkorolev replied:
:+1:, good catch, 0983891.

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+SOFTWARE.
+*******************************************************************************/
+
+// `Piece` is an efficient immutable string: a `const char*` pointer along with the length of what it points to.
```
How strong do you feel about the name "Piece"? Why not make it self-descriptive? Why use the word that means so many different things, including "firearm", "musical composition", and "slice"?

dkorolev replied:
Not strongly. Just need a short and simple name.

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+
+// `Piece` is an efficient immutable string: a `const char*` pointer along with the length of what it points to.
+//
+// It is guaranteed that the actual allocated buffer size is at least one byte more than the reported length,
```
...longer than the reported length by at least one byte...

dkorolev replied:
:+1: , fixed in d7efa31

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+// their underlying `const char*`-s testable for equality. Long live and rest in piece, C+=, we'll miss you.
+//
+// Unlike `UniquePiece`, `Piece` itself does not expose comparison operators. This is done on purpose to make it
+// harder for a culturally uneducated developer to confuse a member of a liberal `Piece` community, who are all
```
What's the upside of dinging "culturally uneducated" developers? Sounds juvenile. Same for ew wordplay ("liberal Piece community"). Totally unnecessary IMHO.

dkorolev replied:
:+1: , fixed in d7efa31

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+// harder for a culturally uneducated developer to confuse a member of a liberal `Piece` community, who are all
+// prouly different and resist exposing any [efficient] way of ordering or even comparing one to another,
+// with a member of an organized `UniquePiece` society, members of which have gladly accepted the deprivation
+// of the right to express their opinion on whether the order should be lexicographical, as long as it is total.
```
total? what does this mean? I'd rewrite this in a simple, straightforward language, free of BBT-style geekicisms.

dkorolev replied:
http://en.wikipedia.org/wiki/Total_order

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+//
+// (Culturally educated developers are hereby granted permission to use `reinterpret_cast<>`. You're welcome.)
+//
+// `PieceDB` is a storage of `UniquePiece`-s, that makes equal `Piece`-s equal, like Samuel Colt did --
```
Here we go, another unfortunate pun...

dkorolev replied:
:+1: , fixed in d7efa31

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+//
+// Both `Piece` and `UniquePiece` are naturally unsafe and require memory for their storage to stay allocated,
+// because every decent architect knows well what happens if `free()` is invoked a bit too prematurely in the
+// evolutionary process of a lifetime of an object. (If you don't believe in evolution, don't use this code.)
```
Who is this for? A good engineer will, at best, cringe, and a bad one will curse.

dkorolev replied:
:+1: , fixed in d7efa31

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+// Note 1: It's not impossible for the order of pointed to strings to actually be lexicographical. However,
+//         in most cases, making it so would be quite a large amount of unnecessary work. C+=. Never forget.
+//
+// Note 2: To make it explicit, since some readers, who totally should, might not reach this point.
```
How about simply "To make it explicit"?

A subtle joke once in a while is ok but it has to be a) subtle and b) rare.

dkorolev replied:
:+1: , fixed in d7efa31 (and shortened the rest)

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+//         Thus, non-metaphorically, the description above makes total sense. Keep in mind though, that
+//         making sense did not help C+= survive. At same same time, although contradictory to most predominant
+//         dogmas, I personally stick with the belief that not only it's important for the things to make sense,
+//         but that making sense is the very only property of things that makes them important. -- D.K.
```
Some people will appreciate this sprinkling of personal philosophy everywhere. Most won't IMHO.

dkorolev replied:
:+1: , removed in d7efa31

znrg commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+//         dogmas, I personally stick with the belief that not only it's important for the things to make sense,
+//         but that making sense is the very only property of things that makes them important. -- D.K.
+//
+// Note 3: `Piece` is a better name than `SlaveString`, right?
```
No. Not right. Equally bad.

dkorolev replied:
Suggestions welcome.

mzhurovich commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+
+struct Piece {
+  const char* S;
+  size_t N;
```
Maybe make the above members `private`?

dkorolev replied:
:+1: , fixed in d7efa31

mzhurovich commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,201 @@
+    N = 0u;
+  }
+
+  void CheckPrivilege(size_t i) const { assert(i <= N); }
```
Why do you need this unobviously named method?

dkorolev commented:
Gentlemen, thank you for the review.

d7efa31 addresses most all of your comments and better name suggestions are welcome.

@mzhurovich @znrg, please take another look.

Thanks,
Dima

mzhurovich commented on Bricks/strings/piece.h:
```diff
@@ -0,0 +1,199 @@
+// their underlying `const char*`-s testable for equality. Long live and rest in piece, C+=, we'll miss you.
+//
+// Unlike `UniquePiece`, `Piece` itself does not expose comparison operators. This is done on purpose to make it
+// immossible to accidentally confuse a member of a liberal `Piece` community, who are all prouly different
```
im**p**ossible

dkorolev commented:
Performance numbers:

```
dima@dima-i9: ~/github/dkorolev/Current/CompactTSV(pride) $ ./.current/test --benchmark=1 --rows=100000 --cols=200
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from CompactTSV
[ RUN      ] CompactTSV.Smoke
Generate:                       6298ms.
Pack:                           14531ms.
Unpack into std::ostringstream: 3242ms.
Unpack into std::string-s:      1782ms.
Unpack into UniqueChunk-s:      492ms.
[       OK ] CompactTSV.Smoke (26362 ms)
[----------] 1 test from CompactTSV (26362 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (26362 ms total)
[  PASSED  ] 1 test.
```

Justification: Cube browser for B.T.

Thanks,
Dima

dkorolev commented:
A cleaner test:

```
dima@dima-i9: ~/github/dkorolev/Current/CompactTSV(pride) $ ./.current/test --benchmark=1 --rows=2000000 --cols=200 --scale=0.1
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from CompactTSV
[ RUN      ] CompactTSV.Smoke
Original TSV size:	1200000000b, or	1144MB.
Packed   TSV size:	2181095b, or	2MB.
Generate:                                   50451ms.
Pack:                                       89877ms.
Unpack into std::ostringstream:             19246ms.
Unpack into std::string-s:                  7ms.
Unpack into std::pair<const char*, size_t>: 5ms.
Unpack into UniqueChunk-s:                  6ms.
```

dkorolev commented:
Okay, `--scale=0.1` results in too many identical values, which makes it a bad test, since a) compression is too good, and b) since there are only a few comparisons, `std::string`-s are not copied often and thus are performant enough.

Here's a more real-life test:

```
./.current/test --benchmark=1 --rows=1000000 --cols=200 --scale=0.25
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from CompactTSV
[ RUN      ] CompactTSV.Smoke
Original TSV size:	600000000b, or	572MB.
Packed   TSV size:	36928890b, or	35MB.
Generate:                                   25236ms.
Pack:                                       45871ms.
Unpack into std::ostringstream:             9283ms.
Unpack into std::string-s:                  358ms.
Unpack into std::pair<const char*, size_t>: 36ms.
Unpack into UniqueChunk-s:                  36ms.
```

dkorolev commented:
The test above was from a ThinkPad T420s, and here's the same one from a MacBook Pro:

```
dima@Users-MacBook-Pro /Users/dima/github/dkorolev/Current/CompactTSV (pride) $ ./.current/test --benchmark=1 --rows=1000000 --cols=200 --scale=0.25
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from CompactTSV
[ RUN      ] CompactTSV.Smoke
Original TSV size:	600000000b, or	572MB.
Packed   TSV size:	36928890b, or	35MB.
Generate:                                   47940ms.
Pack:                                       32943ms.
Unpack into std::ostringstream:             11020ms.
Unpack into std::string-s:                  140ms.
Unpack into std::pair<const char*, size_t>: 24ms.
Unpack into UniqueChunk-s:                  27ms.
[       OK ] CompactTSV.Smoke (92471 ms)
[----------] 1 test from CompactTSV (92471 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (92471 ms total)
[  PASSED  ] 1 test.
```

Bottom line: above 1GB of compressed data per second. My scale.

dkorolev commented:
@mzhurovich, @sompylasar : I've added another method to efficiently test-and-skip prefixes -- need it for the new shiny and now-emerging cube numbers browser that I'm hoping to be done with by tomorrow's BitTorrent meeting.

PTAL! And I suggest we start wrapping up with this pull request until it begins growing again :-)

Thanks,
Dima

sompylasar commented on Bricks/strings/chunk.h:
```diff
@@ -0,0 +1,237 @@
+  // It's `clang-format` inserting those spaces between the operator and closing parenthesis; not me. -- D.K.
+  DEFINE_COMPARATOR(== );
+  DEFINE_COMPARATOR(!= );
+  DEFINE_COMPARATOR(< );
```
@dkorolev  Please could you explain how these `<`, `>`, `<=`, `>=` are going to work with pointers to the string contents (`.S`) which are not guaranteed to be laid in memory in some meaningful sequence, except the ones `Explunge`-d.

dkorolev replied:
They don't have to compare lexicographically -- just form a total order for `std::map`'s red-black tree to work.

sompylasar replied:
Ah, so the returned results don't have to meaningful. OK. A sidenote: it was not obvious to me and makes me first think of some ordering meaningful for the caller, as long as `UniqueChunk` is "exported" to be used standalone from the outside code.

dkorolev replied:
They are meaningful in a sense that an `std::map<>` will do its job, and thus strings can be compared in O(1). The comment in the header mentions this, along with the comparison order of `UniqueChunk`-s being strictly total, yet not lexicographical.

sompylasar replied:
Yes, I now understand, thanks. Although this engineering knowledge is buried under the philosophy and wordplay.

dkorolev replied:
:trollface:

mzhurovich commented on Bricks/strings/chunk.h:
```diff
@@ -0,0 +1,237 @@
+
+// Friendly reminder: `ChunkDB` does not own any strings, it just manages pointers to them. I know right?
+struct ChunkDB {
+  std::unordered_map<Chunk, const UniqueChunk*, Chunk::HashFunction, Chunk::Pride> map;
```
Make it `private`?

dkorolev replied:
:+1: 1142730

mzhurovich commented on Bricks/strings/chunk.h:
```diff
@@ -0,0 +1,239 @@
+// impossible to accidentally confuse a member of a liberal `Chunk` community, who are all proudly different
+// and resist exposing any [efficient] way of ordering or even comparing one to another, with a member of
+// an organized `UniqueChunk` society, members of which have gladly accepted the deprivation
+// of the right to express their opinion on whether the order should be lexicographical, as long as it is total.
```
I'd second @sompylasar that engineering essence is somehow buried in the comments here.
Could you please at least add a link to Wikipedia - `http://en.wikipedia.org/wiki/Total_order`?
Thanks!

dkorolev replied:
Yep! a356e14

mzhurovich commented:
@dkorolev LGTM! :+1:

dkorolev commented:
Yay! Thank you guys!

On Monday, June 8, 2015, Maxim Zhurovich <notifications@github.com> wrote:

> Merged #198 <https://github.com/c5t/Current/pull/198>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/c5t/Current/pull/198#event-325086700>.
>


-- 
And, as always, have a nice day!
Dima

http://dimakorolev.com/ | http://www.linkedin.com/in/dimakorolev |
dmitry.korolev@gmail.com | +1 (312) 593 2783
